examples/fluids/navierstokes.c

```diff
@@ -294,6 +341,91 @@ static PetscErrorCode CreateRestrictionFromPlex(Ceed ceed, DM dm, CeedInt P,
+  ierr = DMGetDimension(dm, &dim); CHKERRQ(ierr);
+  ncompx = dim;
+  dim -= height;
+  Qdim = CeedIntPow(Q, dim);
+  ierr = DMGetCoordinateDM(dm, &dmcoord); CHKERRQ(ierr);
+  ierr = DMPlexSetClosurePermutationTensor(dmcoord, PETSC_DETERMINE, NULL);
+  CHKERRQ(ierr);
+  ierr = CreateRestrictionFromPlex(ceed, dm, P, height, domainLabel, value, restrictq);
+  CHKERRQ(ierr);
+  ierr = CreateRestrictionFromPlex(ceed, dmcoord, 2, height, domainLabel, value, restrictx);
+  CHKERRQ(ierr);
+  CeedElemRestrictionGetNumElements(*restrictq, &localNelem);
+  CeedElemRestrictionCreateStrided(ceed, localNelem, Qdim,
+                                   qdatasize, qdatasize*localNelem*Qdim,
+                                   CEED_STRIDES_BACKEND, restrictqdi);
+  PetscFunctionReturn(0);
+}
+
+// Utility function to create CEED Operator for each Face
+static PetscErrorCode CreateOperatorForFaces(Ceed ceed, DM dm, CeedOperator op_applyVol,
+    CeedQFunction qf_applySur, CeedQFunction qf_setupSur, CeedInt height, PetscInt nFace,
+    PetscInt value[6], CeedInt numP_Sur, CeedInt numQ_Sur, CeedInt qdatasizeSur, CeedInt NqptsSur,
+    CeedBasis basisxSur, CeedBasis basisqSur, CeedVector xcorners, CeedOperator *op_apply) {
+
+  CeedInt dimSur;
+  CeedElemRestriction restrictxSur[6], restrictqSur[6], restrictqdiSur[6];
+  PetscInt dim, localNelemSur[6];
+  CeedVector qdataSur[6];
+  CeedOperator op_setupSur[6], op_applySur[6];
+  DMLabel domainLabel;
+  PetscErrorCode ierr;
+
+  PetscFunctionBeginUser;
+  ierr = DMGetDimension(dm, &dim); CHKERRQ(ierr);
+  dimSur = dim - height;
+  // Composite Operaters
+  CeedCompositeOperatorCreate(ceed, op_apply);
+  CeedCompositeOperatorAddSub(*op_apply, op_applyVol);
+
+  if (nFace){
+      // Get Label for the Faces
+      ierr = DMGetLabel(dm, "Face Sets", &domainLabel); CHKERRQ(ierr);
+      // Create CEED Operator for each Face
+      for(CeedInt i=0; i<nFace; i++){
+        ierr = GetRestrictionForDomain(ceed, dm, height, domainLabel, value[i], numP_Sur,
+                                       numQ_Sur, qdatasizeSur, &restrictqSur[i], &restrictxSur[i],
+                                       &restrictqdiSur[i]); CHKERRQ(ierr);
+        // Create the CEED vectors that will be needed in setup
+        CeedElemRestrictionGetNumElements(restrictqSur[i], &localNelemSur[i]);
+        CeedVectorCreate(ceed, qdatasizeSur*localNelemSur[i]*NqptsSur, &qdataSur[i]);
+
+        // Create the operator that builds the quadrature data for the NS operator
+        CeedOperatorCreate(ceed, qf_setupSur, NULL, NULL, &op_setupSur[i]);
+        CeedOperatorSetField(op_setupSur[i], "dx", restrictxSur[i], basisxSur, CEED_VECTOR_ACTIVE);
+        CeedOperatorSetField(op_setupSur[i], "weight", CEED_ELEMRESTRICTION_NONE,
+                             basisxSur, CEED_VECTOR_NONE);
+        CeedOperatorSetField(op_setupSur[i], "qdataSur", restrictqdiSur[i],
+                             CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
```

Comment:
It looks like `op_setupSur` is never applied, and this is why the setup data is missing/zero when used in `Advection2d_Sur`.

Reply:
Hmm! So, the problem that you previously mentioned still exists.
For applying composite operators, I had followed `t521`. Only, I didn't create any composite operators for the setup  which I think makes sense because the setup QFunctions are not communicating with each other. That is not the issue, or is it?

Reply:
You don't need to make a composite setup operator (I mainly made a composite setup operator for t521 because I wanted to test composite operators with passive outputs). One way or another though, every setup operator needs to be run once. In the case of t521, I bundled them into a singe operator that was applied once, but if there is no reason to compose the operators, then they'll need to be applied individually, like in line 1271 of navierstokes.c on master.

Reply:
Ah, I see now. Then, `op_setupSur` is truly never applied :D 
Thanks, @jedbrown and @jeremylt.